src/backend/distributed/executor/executor.c

```diff
@@ -993,21 +1020,96 @@ FindOrCreateWorkerSession(WorkerPool *workerPool, MultiConnection *connection)
 
 
 /*
- * RunDistributedExecution runs a distributed execution to completion. It
- * creates a wait event set to listen for events on any of the connections
- * and runs the connection state machine when a connection has an event.
+ * ShouldRunTasksSequentially returns true if each of the individual tasks
+ * should be executed one by one. Note that this is different than
+ * MultiShardConnectionType == SEQUENTIAL_CONNECTION case. In that case,
+ * running the tasks across the nodes in parallel is acceptable and implemented
+ * in that way.
+ *
+ * However, the executions that are qualified here would perform poorly if the
+ * tasks across the workers are executed in parallel. We currently qualify only
+ * one class of distributed queries here, multi-row INSERTs. If we do not enforce
+ * true sequential execution, concurrent multi-row upserts could easily form
+ * a distributed deadlock when the upserts touch the same rows.
+ */
+static bool
+ShouldRunTasksSequentially(List *taskList)
+{
+	Task *initialTask = NULL;
+
+	if (list_length(taskList) < 2)
+	{
+		/* single task plans are already qualified as sequential by definition */
+		return false;
+	}
+
+	/* all the tasks are the same, so we only look one */
+	initialTask = (Task *) linitial(taskList);
+	if (initialTask->rowValuesLists != NIL)
+	{
+		/* found a multi-row INSERT */
+		return true;
+	}
+
+	return false;
+}
+
+
+/*
+ * SequentialRunDistributedExecution gets a distributed execution and
+ * executes each individual task in the exection sequentially, one
+ * task at a time. See related function ShouldRunTasksSequentially()
+ * for more detail on the definition of SequentialRun.
+ */
+static void
+SequentialRunDistributedExecution(DistributedExecution *execution)
+{
+	List *taskList = execution->tasksToExecute;
+
+	ListCell *taskCell = NULL;
+
+	foreach(taskCell, taskList)
+	{
+		Task *taskToExecute = (Task *) lfirst(taskCell);
+
+		/* we'll only need one connection */
+		execution->targetPoolSize = 1;
+
+		/* execute each task one by one */
+		execution->tasksToExecute = list_make1(taskToExecute);
+		execution->totalTaskCount = 1;
+		execution->unfinishedTaskCount = 1;
+
+		/* simply call the regular execution function */
+		RunDistributedExecution(execution);
```

Comment:
@marcocitus Does that fit what into what you had in mind?

Reply:
This definitely seems a lot cleaner, could you rebase on unified_executor to make the changes easier to read?

Reply:
should be ready to have a look.